packages/firestore/src/local/web_storage.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code, FirestoreError } from '../util/error';
+import { VisibilityState } from '../core/types';
+import { assert } from '../util/assert';
+import { AsyncQueue } from '../util/async_queue';
+import { debug } from '../util/log';
+
+/**
+ * Refresh the contents of WebStorage every four seconds.
+ */
+const WEB_STORAGE_REFRESH_INTERVAL_MS: number = 4000;
+
+// Prefix keys used in WebStorage.
+const VISIBILITY_PREFIX = 'visibility';
+
+const LOG_TAG = 'WebStorage';
+
+/**
+ * WebStorage of the Firestore client. Firestore uses WebStorage for cross-tab
+ * notifications and to persist the metadata state of each tab. WebStorage is
+ * used to perform leader election and to inform other tabs about changes in the
+ * IndexedDB-backed persistence layer.
+ *
+ * WebStorage is only used in Persistence-enabled Firestore instances. If you
+ * are not using Persistence, consider using the `NoOpWebStorage` class.
+ */
+export interface WebStorage {
```

Comment:
I'm not wild about this name--it doesn't really describe what this thing is for. Also, we already have LocalStore and RemoteStore, so if this thing really is going to be named after storing things we should probably go for a name ending in "Store".

However, the first sentence of the comment indicates this is cross tab notifications. How about `TabNotificationChannel`?

Reply:
I explicitly avoided naming anything LocalStorage in this code since this will create mighty confusion with LocalStore. `TabNotificationChannel` works for me and PR has been updated.